In [86]:
import pandas as pd
import numpy as np
import re
import warnings
warnings.filterwarnings('ignore')


## Python Data Structures
### 1- Lists


In [87]:
squares_list = [0,1,4,9,16,25]
squares_list

[0, 1, 4, 9, 16, 25]

In [88]:
# Indexing returns the item 
squares_list[2]

4

In [89]:
# Slicing retruns new list 
squares_list[2:4]

[4, 9]

In [90]:
# Negative indexing returns from the last 
squares_list[-2]

16

In [91]:
#Item assignment 
squares_list[2] = 3
squares_list

[0, 1, 3, 9, 16, 25]

### 2- Strings


In [92]:
greeting = 'Hello'

print ("First Character of the string is ",greeting[0])
print ("The length of the string= ",len(greeting))
print ("We can add strings with + like: ", greeting+" World")

First Character of the string is  H
The length of the string=  5
We can add strings with + like:  Hello World


In [93]:
# Strings can also be sliced 

greeting[1:]

'ello'

In [94]:
# Raw string used to pass the string as it is 
# Note the Difference 

normal_string = '\n is a new line character by default'
raw_string = r'\n is a new line character by default'
print (raw_string,normal_string)

\n is a new line character by default 
 is a new line character by default


In [97]:
# We can't replace characters in strings with indexing 
# The following line should produce error
# We will try to replace "Hello" with "Hi"

greeting[1:] = 'i'

TypeError: 'str' object does not support item assignment

In [98]:
greeting.replace('Hello', 'Hi')

'Hi'

### 3- Tuples

In [99]:
tuple_example = 1,2,3,4,5
tuple_example

(1, 2, 3, 4, 5)

In [100]:
# Indexing 
tuple_example[2]

3

In [101]:
#Slicing 
tuple_example[1:3]

(2, 3)

In [102]:
# NO Item assignment
# This line should produce error

tuple_example[2] = 3

TypeError: 'tuple' object does not support item assignment

### 4- Dictionary

In [103]:
employees = {'Hussien': 9073 , 'Ahmed' : 9182 , 'Sameh': 9223, 'Salma': 9330}
employees

{'Hussien': 9073, 'Ahmed': 9182, 'Sameh': 9223, 'Salma': 9330}

In [104]:
# Accessing by key
employees['Ahmed']

9182

In [105]:
employees.keys()

dict_keys(['Hussien', 'Ahmed', 'Sameh', 'Salma'])

In [106]:
employees.values()

dict_values([9073, 9182, 9223, 9330])

### 5- Matrix

In [107]:
import numpy as np
import pandas as pd
mat = np.matrix([[1,2,3],[4,5,6]])

In [108]:
mat

matrix([[1, 2, 3],
        [4, 5, 6]])

Broadcasting

In [109]:
mat + 1

matrix([[2, 3, 4],
        [5, 6, 7]])

In [110]:
mat * 3

matrix([[ 3,  6,  9],
        [12, 15, 18]])

Matrix Multiplication

In [111]:
a = np.matrix([[1,2,3],[4,5,6]])
b = np.matrix([[7,8],[9,10],[11,12]])

In [112]:
a.dot(b)

matrix([[ 58,  64],
        [139, 154]])

# 6- Data Frames

In [113]:
df = pd.read_csv("train.csv")

In [114]:
df.tail(10)

Loan_ID  Gender Married Dependents     Education Self_Employed  \
604  LP002959  Female     Yes          1      Graduate            No   
605  LP002960    Male     Yes          0  Not Graduate            No   
606  LP002961    Male     Yes          1      Graduate            No   
607  LP002964    Male     Yes          2  Not Graduate            No   
608  LP002974    Male     Yes          0      Graduate            No   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
604            12000                0.0       496.0             360.0   
605             2400             3800.0         NaN             180.0   
606             3400             2500.0       173.0             360.0   
607             3987             1411.0       157.0             360.0   
608             3232             1950.0       108.0             360.0   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
604             1.0     Semiurban           Y  
605             1.0         Urban           N  
606             1.0     Semiurban           Y  
607             1.0         Rural           Y  
608             1.0         Rural           Y  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N

In [115]:
df.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [116]:
df['Loan_Amount_Term'].head(2)

0    360.0
1    360.0
Name: Loan_Amount_Term, dtype: float64

In [117]:
df.Gender.value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

In [118]:
pd.crosstab(df.Gender, df.Married)

Married   No  Yes
Gender           
Female    80   31
Male     130  357

In [119]:
df.Education.value_counts()

Graduate        480
Not Graduate    134
Name: Education, dtype: int64

## Python Iteration and Conditional Constructs
### 1-Loops

In [120]:
# Get the factorial 
fact = 1
N = 5

for i in range (1,N+1):
    fact *= i

fact 

120

In [121]:
#importing library
import math 

# Loop on items
for i in squares_list:
      print (i,"The root = ",np.sqrt(i))

0 The root =  0.0
1 The root =  1.0
3 The root =  1.7320508075688772
9 The root =  3.0
16 The root =  4.0
25 The root =  5.0


In [122]:
# Conditions 
N = 4

if (N%2==0):
    print("Even")
else:
    print ("Odd")

Even


Arabic and English Text Cleaning

In [123]:
import nltk

def replaceMultiple(mainString, toBeReplaces, newString):
    # Iterate over the strings to be replaced
    for elem in toBeReplaces :
        # Check if string is in the main string
        if elem in mainString :
            # Replace the string
            mainString = mainString.replace(elem, newString)
    
    return  mainString

def remove_stop_words(text):
    filtered_word_list = [] #make a copy of the word_list
    for word in text.split(): # iterate over word_list
        if((word not in english_stopwords) and (word not in arabic_stopwords)):
            filtered_word_list.append(word)# remove word from filtered_word_list if it is a stopword
    return filtered_word_list



#nltk.download('stopwords')
english_stopwords = list(set(nltk.corpus.stopwords.words("english")))
arabic_stopwords = list(set(nltk.corpus.stopwords.words("arabic")))

In [124]:
df = pd.read_csv('BeautySegment.csv', encoding='UTF-8')

In [125]:
df.head()

Topics              author author country             domain  \
0     NaN  nadyah_photography  United States  www.instagram.com   
1     NaN  nadyah_photography  United States  www.instagram.com   
2     NaN  nadyah_photography  United States  www.instagram.com   
3     NaN           alkhateeb   Saudi Arabia        twitter.com   
4     NaN      ShahadAlAngari   Saudi Arabia        twitter.com   

                                           full text   gender interest  \
0  - The : @a_umshahad ✨ - a place : @lastoriacaf...  unknown       []   
1  - The : @a_umshahad ✨ - a place : @lastoriacaf...  unknown       []   
2  - The : @a_umshahad ✨ - a place : @lastoriacaf...  unknown       []   
3  RT @ArtByLaRoque: There isn't beauty more fulf...  unknown       []   
4                     There is beauty in simplicity.  unknown       []   

                                                 url             Subject  
0           https://www.instagram.com/p/BiNsPKilwoV/  Beauty Enthusiasts  
1           https://www.instagram.com/p/BiNtTZBlaf7/  Beauty Enthusiasts  
2           https://www.instagram.com/p/BiNuOeUln-N/  Beauty Enthusiasts  
3  http://twitter.com/alkhateeb/statuses/99111003...  Beauty Enthusiasts  
4  http://twitter.com/ShahadAlAngari/statuses/991...  Beauty Enthusiasts

In [126]:
snippets = df['full text'].astype(str)

In [127]:
snippets[33]

'RT @vencaresa: VENCARE شامبو & لوشن & أمبولات متوفرة على مواقع البيع الألكتروني ( سوق كوم & وادي & إي مول & كوبون % خصوماتي & دلوني & سيدتي مول ) وفي الصيدليات المتخصصة في التجميل pic.twitter.com/xGSwyKh6ss'

List Comprehension

In [128]:
snippets = [' '.join(remove_stop_words(x)) for x in snippets]

In [129]:
snippets[33]

'RT @vencaresa: VENCARE شامبو & لوشن & أمبولات متوفرة مواقع البيع الألكتروني ( سوق كوم & وادي & مول & كوبون % خصوماتي & دلوني & سيدتي مول ) وفي الصيدليات المتخصصة التجميل pic.twitter.com/xGSwyKh6ss'

In [130]:
sn = snippets[33]

In [131]:
# remove numbers
''.join([i for i in sn if not i.isdigit()])


'RT @vencaresa: VENCARE شامبو & لوشن & أمبولات متوفرة مواقع البيع الألكتروني ( سوق كوم & وادي & مول & كوبون % خصوماتي & دلوني & سيدتي مول ) وفي الصيدليات المتخصصة التجميل pic.twitter.com/xGSwyKhss'

In [132]:
import re
#remove hashtags, links, attached pictures, and RT @
re.sub(r"(?:@\S*|#\S*|http(?=.*://)\S*|pic.twitter\S*|RT(?=.*@)\S*)", "", sn)


'  VENCARE شامبو & لوشن & أمبولات متوفرة مواقع البيع الألكتروني ( سوق كوم & وادي & مول & كوبون % خصوماتي & دلوني & سيدتي مول ) وفي الصيدليات المتخصصة التجميل '

In [133]:
'''
Replace multiple characters / strings from a string
'''
# Replace all the occurrences of string in list by AA in the main list 
otherStr = replaceMultiple(sn.lower(), ['%','&',':', '.',',', '?', ',', '،', '؛', '‘', '’', '`', '~', '?', '؟', '-', ';', 'ـ', '#', '_'] , " ")
otherStr = replaceMultiple(otherStr, ['ّ', 'َ', 'ً', 'ُ', 'ٌ', 'ِ', 'ٍ', 'ْ'], "")
otherStr = replaceMultiple(otherStr, ['أ', 'آ', 'إ'], 'ا')
otherStr = replaceMultiple(otherStr, ['ى'], "ي")
otherStr = replaceMultiple(otherStr, ['ة'], 'ه')
otherStr = replaceMultiple(otherStr, ['(', ')', '{', '}', '[', ']', '<', '>', '<', '>', 'ـ', '{','}','–', '.', ':',','], '')
otherStr

'rt @vencaresa  vencare شامبو   لوشن   امبولات متوفره مواقع البيع الالكتروني  سوق كوم   وادي   مول   كوبون   خصوماتي   دلوني   سيدتي مول  وفي الصيدليات المتخصصه التجميل pic twitter com/xgswykh6ss'

Let's do the whole thing again and compare

In [134]:
sn 

'RT @vencaresa: VENCARE شامبو & لوشن & أمبولات متوفرة مواقع البيع الألكتروني ( سوق كوم & وادي & مول & كوبون % خصوماتي & دلوني & سيدتي مول ) وفي الصيدليات المتخصصة التجميل pic.twitter.com/xGSwyKh6ss'

In [135]:
clean_sn = ''.join([i for i in sn if not i.isdigit()])
clean_sn = re.sub(r"(?:@\S*|#\S*|http(?=.*://)\S*|pic.twitter\S*|RT(?=.*@)\S*)", "", clean_sn)
otherStr = replaceMultiple(clean_sn.lower(), ['%','&',':', '.',',', '?', ',', '،', '؛', '‘', '’', '`', '~', '?', '؟', '-', ';', 'ـ', '#', '_'] , " ")
otherStr = replaceMultiple(otherStr, ['ّ', 'َ', 'ً', 'ُ', 'ٌ', 'ِ', 'ٍ', 'ْ'], "")
otherStr = replaceMultiple(otherStr, ['أ', 'آ', 'إ'], 'ا')
otherStr = replaceMultiple(otherStr, ['ى'], "ي")
otherStr = replaceMultiple(otherStr, ['ة'], 'ه')
otherStr = replaceMultiple(otherStr, ['(', ')', '{', '}', '[', ']', '<', '>', '<', '>', 'ـ', '{','}','–', '.', ':',','], '')
clean_sn = otherStr
clean_sn

'  vencare شامبو   لوشن   امبولات متوفره مواقع البيع الالكتروني  سوق كوم   وادي   مول   كوبون   خصوماتي   دلوني   سيدتي مول  وفي الصيدليات المتخصصه التجميل '

# Categorical Variables Representation
One-Hot-Encoding

In [136]:
pd.get_dummies(df['author country'])

Afghanistan  Algeria  Argentina  Armenia  Australia  Austria  Bahrain  \
0                0        0          0        0          0        0        0   
1                0        0          0        0          0        0        0   
2                0        0          0        0          0        0        0   
3                0        0          0        0          0        0        0   
4                0        0          0        0          0        0        0   
5                0        0          0        0          0        0        0   
6                0        0          0        0          0        0        0   
7                0        0          0        0          0        0        0   
8                0        0          0        0          0        0        0   
9                0        0          0        0          0        0        0   
10               0        0          0        0          0        0        0   
11               0        0          0        0          0        0        0   
12               0        0          0        0          0        0        0   
13               0        0          0        0          0        0        0   
14               0        0          0        0          0        0        0   
15               0        0          0        0          0        0        0   
16               0        0          0        0          0        0        0   
17               0        0          0        0          0        0        0   
18               0        0          0        0          0        0        0   
19               0        0          0        0          0        0        0   
20               0        0          0        0          0        0        0   
21               0        0          0        0          0        0        0   
22               0        0          0        0          0        0        0   
23               0        0          0        0          0        0        0   
24               0        0          0        0          0        0        0   
25               0        0          0        0          0        0        0   
26               0        0          0        0          0        0        0   
27               0        0          0        0          0        0        0   
28               0        0          0        0          0        0        0   
29               0        0          0        0          0        0        0   
...            ...      ...        ...      ...        ...      ...      ...   
11323            0        0          0        0          0        0        0   
11324            0        0          0        0          0        0        0   
11325            0        0          0        0          0        0        0   
11326            0        0          0        0          0        0        0   
11327            0        0          0        0          0        0        0   
11328            0        0          0        0          0        0        0   
11329            0        0          0        0          0        0        0   
11330            0        0          0        0          0        0        0   
11331            0        0          0        0          0        0        0   
11332            0        0          0        0          0        0        0   
11333            0        0          0        0          0        0        0   
11334            0        0          0        0          0        0        0   
11335            0        0          0        0          0        0        0   
11336            0        0          0        0          0        0        0   
11337            0        0          0        0          0        0        0   
11338            0        0          0        0          0        0        0   
11339            0        0          0        0          0        0        0   
11340            0        0          0        0          0        0        0   
11341  

In [137]:
pd.get_dummies(df['author country'])[['United Arab Emirates', 'United States', 'Saudi Arabia']].head(10)

United Arab Emirates  United States  Saudi Arabia
0                     0              1             0
1                     0              1             0
2                     0              1             0
3                     0              0             1
4                     0              0             1
5                     0              0             1
6                     0              1             0
7                     0              0             1
8                     0              0             1
9                     0              1             0

##  Exploratory analysis in Python using "Pandas"

In [138]:
# Importing libraries
import pandas as pd 
import numpy as np
import matplotlib as plt

%matplotlib inline

 **VARIABLE DESCRIPTIONS:**


**Loan_ID**	  :                Unique Loan ID

**Gender**	    :              Male/ Female

**Married**	    :              Applicant married (Y/N)

**Dependents**	   :       Number of dependents

**Education**	    :      Applicant Education (Graduate/ Under Graduate)

**Self_Employed**	   :       Self employed (Y/N)

**ApplicantIncome**	  :        Applicant income

**CoapplicantIncome**:	  Coapplicant income

**LoanAmount**	      :    Loan amount in thousands

**Loan_Amount_Term**	:  Term of loan in months

**Credit_History**	   :       credit history meets guidelines

**Property_Area**	   :       Urban/ Semi Urban/ Rural

**Loan_Status**	     :     Loan approved (Y/N)

In [139]:
df = pd.read_csv("train.csv")

In [140]:
# Quick data exploration
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [141]:
# Data frame dimensions
df.shape

(614, 13)

In [142]:
# Get statistics about all numerical columns
df.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

### Note
There are some columns that count is less than 614 rows 
Which means that they have "missing values"

In [143]:
# Exploring non-numerical columns

df['Property_Area'].value_counts()

Semiurban    233
Urban        202
Rural        179
Name: Property_Area, dtype: int64

### Dealing with missing values

In [144]:
# Finding how number of missing values per column

len(df) - df.count()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [145]:
## Filling missing values with the mean(average)
df['LoanAmount'].fillna(df['LoanAmount'].mean(), inplace=True)


In [146]:
## Filling missing values with the mode (most frequent value)
## look at 'Self_Employed' column 
df['Self_Employed'].value_counts()

No     500
Yes     82
Name: Self_Employed, dtype: int64

In [147]:
## Replacing misssing values with "No" will have high probability of success 
df['Self_Employed'].fillna('No',inplace=True)

## Building a Predictive Model in Python


In [148]:
## Dealing with all missing values in columns

df['Gender'].fillna(df['Gender'].mode()[0], inplace=True)
df['Married'].fillna(df['Married'].mode()[0], inplace=True)
df['Dependents'].fillna(df['Dependents'].mode()[0], inplace=True)
df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].mode()[0], inplace=True)
df['Credit_History'].fillna(df['Credit_History'].mode()[0], inplace=True)

#### - Labeling Categorical Variables

In [149]:
# Labeling categorical variables

from sklearn.preprocessing import LabelEncoder
var_mod = ['Gender','Married','Dependents','Education','Self_Employed','Property_Area','Loan_Status'] # Column Names 

le = LabelEncoder() # Creating the enconder

for i in var_mod: #Looping on columns
    df[i] = le.fit_transform(df[i]) # encoding each column
df.dtypes 

Loan_ID               object
Gender                 int64
Married                int64
Dependents             int64
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

In [150]:
df.head(10)

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0  LP001002       1        0           0          0              0   
1  LP001003       1        1           1          0              0   
2  LP001005       1        1           0          0              1   
3  LP001006       1        1           0          1              0   
4  LP001008       1        0           0          0              0   
5  LP001011       1        1           2          0              1   
6  LP001013       1        1           0          1              0   
7  LP001014       1        1           3          0              0   
8  LP001018       1        1           2          0              0   
9  LP001020       1        1           1          0              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0  146.412162             360.0   
1             4583             1508.0  128.000000             360.0   
2             3000                0.0   66.000000             360.0   
3             2583             2358.0  120.000000             360.0   
4             6000                0.0  141.000000             360.0   
5             5417             4196.0  267.000000             360.0   
6             2333             1516.0   95.000000             360.0   
7             3036             2504.0  158.000000             360.0   
8             4006             1526.0  168.000000             360.0   
9            12841            10968.0  349.000000             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              0            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1  
5             1.0              2            1  
6             1.0              2            1  
7             0.0              1            0  
8             1.0              2            1  
9             1.0              1            0

#### - Splitting Data 

In [151]:
from sklearn.model_selection import train_test_split
predictor_var = ['Credit_History','Education','Married','Self_Employed','Property_Area']
outcome_var = 'Loan_Status'

X_train, X_test, y_train, y_test = train_test_split(df[predictor_var], df[outcome_var], test_size=0.33, random_state=42)

#### - Fitting the model to training data (base model)

In [152]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,y_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

#### - Predict the labels for test data

In [153]:
y_pred = model.predict(X_test)
y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 0, 1, 1], dtype=int64)

#### - Evaluating the model

In [154]:
from sklearn.metrics import accuracy_score

print('Accuracy: %.2f%%' % (accuracy_score(y_test, y_pred) * 100))

Accuracy: 79.80%


## Model Selection

In [155]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

models = [('LR', LogisticRegression()),
         ('KNN', KNeighborsClassifier()),
         ('CART', DecisionTreeClassifier()),
         ('NB', GaussianNB()),
         ('SVM', SVC()),
         ('RF', RandomForestClassifier())]
seed = 1073
results = []
names = []
scoring = 'accuracy'
X = X_train
Y = y_train
for name, model in models:
    kfold = KFold(n_splits=5, random_state=seed)
    cv_results = cross_val_score(model, X, Y, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
    print(msg)

LR: 0.815075 (0.016269)
KNN: 0.788334 (0.024979)
CART: 0.790685 (0.024158)
NB: 0.815075 (0.016269)
SVM: 0.815075 (0.016269)
RF: 0.790714 (0.019884)


## Hyperparameter Tuning

In [156]:
import pprint
from sklearn.model_selection import RandomizedSearchCV
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 50, stop = 250, num = 5)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 7)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
random_grid

{'n_estimators': [50, 100, 150, 200, 250],
 'max_features': ['auto', 'sqrt'],
 'max_depth': [10, 26, 43, 60, 76, 93, 110, None],
 'min_samples_split': [2, 5, 10],
 'min_samples_leaf': [1, 2, 4],
 'bootstrap': [True, False]}

In [157]:
# Use the random grid to search for best hyperparameters
# First create the base model to tune
rf = RandomForestClassifier()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_random = RandomizedSearchCV(estimator = rf, param_distributions = random_grid, n_iter = 100, cv = 5, verbose=2, random_state=42, n_jobs = -1)
# Fit the random search model
rf_random.fit(X, Y)

print( rf_random.best_estimator_ )
print( rf_random.best_score_ )
print( rf_random.best_params_ )

Fitting 5 folds for each of 100 candidates, totalling 500 fits


[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:   16.1s
[Parallel(n_jobs=-1)]: Done 154 tasks      | elapsed:   31.7s
[Parallel(n_jobs=-1)]: Done 357 tasks      | elapsed:  1.0min
[Parallel(n_jobs=-1)]: Done 500 out of 500 | elapsed:  1.3min finished


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=26, max_features='sqrt', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)
0.8150851581508516
{'n_estimators': 200, 'min_samples_split': 2, 'min_samples_leaf': 4, 'max_features': 'sqrt', 'max_depth': 26, 'bootstrap': True}


In [158]:
clf = RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=26, max_features='sqrt', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [159]:
clf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=26, max_features='sqrt', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)